dotnet/src/Connectors/Connectors.GoogleVertexAI/Core/Gemini/GeminiMetadata.cs

```diff
@@ -0,0 +1,117 @@
+        internal init => this.SetValueInDictionary(value);
+    }
+
+    /// <summary>
+    /// The count of tokens in the prompt.
+    /// </summary>
+    public int PromptTokenCount
+    {
+        get => (this.GetValueFromDictionary() as int?) ?? 0;
+        internal init => this.SetValueInDictionary(value);
+    }
+
+    /// <summary>
+    /// The count of token in the current candidate.
+    /// </summary>
+    public int CurrentCandidateTokenCount
+    {
+        get => (this.GetValueFromDictionary() as int?) ?? 0;
+        internal init => this.SetValueInDictionary(value);
+    }
+
+    /// <summary>
+    /// The total count of tokens of the all candidate responses.
+    /// </summary>
+    public int CandidatesTokenCount
+    {
+        get => (this.GetValueFromDictionary() as int?) ?? 0;
+        internal init => this.SetValueInDictionary(value);
+    }
+
+    /// <summary>
+    /// The total count of tokens (prompt + total candidates token count).
+    /// </summary>
+    public int TotalTokenCount
+    {
+        get => (this.GetValueFromDictionary() as int?) ?? 0;
+        internal init => this.SetValueInDictionary(value);
+    }
+
+    /// <summary>
+    /// The reason why prompt was blocked.
+    /// </summary>
+    public string? PromptFeedbackBlockReason
+    {
+        get => this.GetValueFromDictionary() as string;
+        internal init => this.SetValueInDictionary(value);
+    }
+
+    /// <summary>
+    /// List of safety ratings for the prompt feedback.
+    /// </summary>
+    public IReadOnlyList<GeminiSafetyRating>? PromptFeedbackSafetyRatings
+    {
+        get => this.GetValueFromDictionary() as IReadOnlyList<GeminiSafetyRating>;
+        internal init => this.SetValueInDictionary(value);
+    }
+
+    /// <summary>
+    /// List of safety ratings for the response.
+    /// </summary>
+    public IReadOnlyList<GeminiSafetyRating>? ResponseSafetyRatings
+    {
+        get => this.GetValueFromDictionary() as IReadOnlyList<GeminiSafetyRating>;
+        internal init => this.SetValueInDictionary(value);
+    }
+
+    /// <summary>
+    /// Converts a dictionary to a <see cref="GeminiMetadata"/> object.
+    /// </summary>
+    public static GeminiMetadata FromDictionary(IReadOnlyDictionary<string, object?> dictionary) => dictionary switch
+    {
+        null => throw new ArgumentNullException(nameof(dictionary)),
+        GeminiMetadata metadata => metadata,
+        IDictionary<string, object?> metadata => new GeminiMetadata(metadata),
+        _ => new GeminiMetadata(dictionary.ToDictionary(pair => pair.Key, pair => pair.Value))
+    };
+
+    private void SetValueInDictionary(object? value, [CallerMemberName] string propertyName = "")
+        => this.Dictionary[propertyName] = value;
+
+    private object? GetValueFromDictionary([CallerMemberName] string propertyName = "")
+        => this.Dictionary.TryGetValue(propertyName, out var value) ? value : null;
```

Comment:
It will be sufficient if I remove CallerMemberName and pass `nameof(<property>)` as method parameter?

Reply:
Yup! Thanks.

Reply:
@stephentoub Fixed but could you show me sample where CallerMemberName can fail? I am curious, should I avoid using CallerMemberName in code?

Reply:
For example, right now you have this:
```C#
    public string? PromptFeedbackBlockReason
    {
        get => this.GetValueFromDictionary() as string;
        internal init => this.SetValueInDictionary(value);
    }
```
That means if metadata was stored into this via PromptFeedbackBlockReason, a consumer could access it as:
```C#
var metadata = ...;
string? reason = metadata["PromptFeedbackBlockReason"];
```
Now imagine someone in the future innocently refactors to this:
```C#
    public string? PromptFeedbackBlockReason
    {
        get => GetStringValueFromDictionary();
        internal init => this.SetStringValueInDictionary(value);
    }

    private string? GetStringValueFromDictionary() => this.GetValueFromDictionary() as string;
    private void SetStringValueInDictionary(string value) => SetValueInDictionary(value);
```
That consumer will now be broken, because the string won't have been stored in metadata["PromptFeedbackBlockReason"] but rather in metadata["SetStringValueInDictionary"].

Or let's say it was refactored to just:
```C#
    public string? PromptFeedbackBlockReason
    {
        get => GetStringValueFromDictionary();
        internal init => this.SetValueInDictionary(value);
    }

    private string? GetStringValueFromDictionary() => this.GetValueFromDictionary() as string;
```
Now access to PromptFeedbackBlockReason will throw because it will be trying to read metadata["GetStringValueFromDictionary"] from the dictionary.

Reply:
I didn't think about that case. Thanks for clarification. I will remember for the future :)